lib/responder.js

```diff
@@ -447,49 +535,31 @@ MultipartDispatcher.prototype.objectStream = function dispatchMultipartObjectStr
 
   var partListener = function objectPartListener(partReadStream) {
     parsingParts++;
+    operation.logger.debug('parsing part %d', parsingParts);
+
     partReadStream.on('header', partHeadersListener);
     partReadStream.on('error',  errorListener);
-    queuedReader.addReader(partReadStream);
+    partReaderQueue.addLast(partReadStream);
+
+    if (partReaderQueue.isLast()) {
+      var partRead = concatStream(partFinisher);
+      partRead.on('error', errorListener);
+      partReadStream.pipe(partRead);
+    }
   };
 
   var parseFinishListener = function objectParseFinishListener() {
     hasParsed = true;
-    doneChecker();
+    responseFinisher();
   };
 
   var responseEndListener = function objectResponseEndListener() {
     hasEnded = true;
-    doneChecker();
+    responseFinisher();
   };
 
-  /**
-   * Check if HTTP response has ended, Dicer has finished parsing,
-   * and the Queue is empty. If all are true, then we can call end()
-   * on output stream.
-   */
-  var doneChecker = function doneChecker() {
-
-    if (queuedReader.isQueueEmpty() && hasParsed && hasEnded) {
-      if (metadataBuffer !== null) {
-        var objectQueue = queuedReader.getItemQueue();
-        operation.queueMetadata(metadataBuffer, objectQueue);
-        metadataBuffer = null;
-      }
-
-      rawHeaderQueue      = null;
-      queuedReader        = null;
-      parser              = null;
-      partHeadersListener = null;
-      partListener        = null;
-      parseFinishListener = null;
-      responseEndListener = null;
-      partTransform       = null;
-
-      operation.outputStream.end();
-    } else if (!hasEnded && parsedParts === parsingParts) {
-      parser.emit('drain');
-    }
-
+  var drainListener = function objectDrainListener() {
+    response.resume();
```

Comment:
Two belated questions

*  Should the drain listener resume only if the response has not ended? 
*  Should the drain listener start a concat stream on the next part?
*  To be more resillent on the previous bullet, should the dispatcher have an additional flag in its state to keep track of whether the first part is waiting or not?

Reply:
Thanks so much for the comments. I addressed the first two bullets with https://github.com/marklogic/node-client-api/pull/422/commits/fdde417708efa010a9c7c5e37daa74acd69aad3d. For the third bullet, seems like existing queue methods are enough to keep track. By "first part," do you mean the first part initially added to the queue or just a part in the queue during a resume step?

Reply:
My concern is whether the drain event might ever be fired when the first part is already concatenating. In principle, that shouldn't happen, but sometimes it's useful to guard against things that shouldn't happen.  If an existing flag already records whether the first part is concatenating, then nothing else is needed.

Reply:
Does setting an isConcatenating flag here address your concerns? https://github.com/marklogic/node-client-api/pull/422/commits/ea415d597944b097123d55a015e9aa1fed5a1834

Reply:
Most definitely.  Thanks.